src/dataflow/stream/write_stream.rs

```diff
@@ -122,46 +125,68 @@ impl<D: Data> WriteStream<D> {
             self.name(),
             self.id()
         );
-        self.stream_closed = true;
+        self.stats.lock().unwrap().close_stream();
         self.pusher = None;
     }
 
     /// Updates the last watermark received on the stream.
     ///
     /// # Arguments
     /// * `msg` - The message to be sent on the stream.
-    fn update_watermark(&mut self, msg: &Message<D>) -> Result<(), WriteStreamError> {
+    fn update_statistics(&mut self, msg: &Message<D>) -> Result<(), WriteStreamError> {
         match msg {
             Message::TimestampedData(td) => {
-                if td.timestamp < self.low_watermark {
+                let mut stats = self.stats.lock().unwrap();
+                if td.timestamp < *stats.low_watermark() {
                     return Err(WriteStreamError::TimestampError);
                 }
+                // Increment the message count.
+                stats
+                    .condition_context
+                    .increment_msg_count(self.id(), td.timestamp.clone())
             }
             Message::Watermark(msg_watermark) => {
-                if msg_watermark < &self.low_watermark {
+                let mut stats = self.stats.lock().unwrap();
+                if msg_watermark < stats.low_watermark() {
                     return Err(WriteStreamError::TimestampError);
                 }
                 slog::debug!(
                     crate::TERMINAL_LOGGER,
                     "Updating watermark on WriteStream {} (ID: {}) from {:?} to {:?}",
                     self.name(),
                     self.id(),
-                    self.low_watermark,
+                    stats.low_watermark(),
                     msg_watermark
                 );
-                self.low_watermark = msg_watermark.clone();
+                stats.update_low_watermark(msg_watermark.clone());
+
+                // Notify the arrival of the watermark.
+                stats
+                    .condition_context
+                    .notify_watermark_arrival(self.id(), msg_watermark.clone());
             }
         }
         Ok(())
     }
+
+    /// Evaluates a condition function on the statistics of the write stream.
+    pub fn evaluate_condition(&self, condition_fn: &CondFn) -> bool {
```

Comment:
Can we also move this to the OperatorExecutor and add a `pub(crate) get_statistics(&self) -> Arc<Mutex<WriteStreamStatistics>>`?

Also, there's a chance of starvation because we're locking an `std::sync::Mutex` blocks the entire thread which can cause problems in an async context. It's fine for now and I'm not sure there's a quick fix, but worth keeping in mind.